src/libraries/System.Net.Quic/ref/System.Net.Quic.cs

```diff
@@ -32,16 +32,28 @@ public sealed partial class QuicConnection : System.IDisposable
         public System.Threading.Tasks.ValueTask<System.Net.Quic.QuicStream> OpenBidirectionalStreamAsync(System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
         public System.Threading.Tasks.ValueTask<System.Net.Quic.QuicStream> OpenUnidirectionalStreamAsync(System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
     }
-    public partial class QuicConnectionAbortedException : System.Net.Quic.QuicException
+    public enum QuicError
     {
-        public QuicConnectionAbortedException(string message, long errorCode) : base(default(string)) { }
-        public long ErrorCode { get { throw null; } }
+        Success = 0,
+        InternalError = 1,
+        ConnectionAborted = 2,
+        StreamAborted = 3,
+        AddressInUse = 4,
+        InvalidAddress = 5,
+        ConnectionTimeout = 6,
+        HostUnreachable = 7,
+        ConnectionRefused = 8,
+        VersionNegotiationError = 9,
+        ConnectionIdle = 10,
+        ProtocolError = 11,
+        OperationAborted = 12,
     }
-    public partial class QuicException : System.Exception
+    public sealed partial class QuicException : System.IO.IOException
     {
-        public QuicException(string? message) { }
-        public QuicException(string? message, System.Exception? innerException) { }
-        public QuicException(string? message, System.Exception? innerException, int result) { }
+        public QuicException(System.Net.Quic.QuicError error, string message, long? applicationErrorCode, System.Exception? innerException) { }
+        public QuicException(System.Net.Quic.QuicError error, string message, long? applicationErrorCode, System.Exception? innerException, int result) { }
```

Comment:
Do we have any precedents for taking `innerException` before other arguments? I'm not aware of that, and to me it feels like there is an implicit rule for placing it at the end.  I wish we had a rule for `message`. The inconsistency in .NET is quite annyoing, see: `ArgumentNullException(paramName, message)` vs. `ArgumentException(message, paramName)`

The `QuicException(QuicError, string, long?, System.Exception?)` variant is logical in a sense that the optional parameters are at the end.

@rzikm the second variant with `result` has not been proposed/approved, and I think we don't need it, since the `Exception.HResult` setter is public.

Reply:
> Do we have any precedents for taking innerException before other arguments? I'm not aware of that, and to me it feels like there is an implicit rule for placing it at the end.

OperationCanceledException, WebException, XmlException, ...

Reply:
I thought they were both code-first, which seemed right to me, because "in an ideal world" I'd expect that the code would be the only required parameter (and the message would be defaulted or keyed off of the code).

Reply:
If we want to do code first for both, that's fine.  The strangeness to me is primarily the `long? applicationErrorCode` between the message and the innerException.

Reply:
Moving the `applicationErrorCode` more to the left is fine by me.

We have places where we use slightly different message based on where we throw the exception, e.g for `OperationAborted`, we have "The operation was aborted" or "Reading has been aborted on this stream" or something like that, so we would definitely need an overload with the message parameter.

I can also add the "simple" overloads @bartonjs mentioned and move the message selection from `ThrowHelper` to the exception ctor if that is the consensus, even if it deviates slightly from the API as was approved.